### PR TITLE
Multiple CI Workflows

### PR DESCRIPTION
**Pull request by DorianGray**
On a test run, this dropped our total CI time from 7m20s to 4m55s an gives us the ability to re-run individual segments.

### CODE REVIEW

**Comment by DorianGray** [comment]
depends on https://github.com/qtile/qtile/pull/2102

**Review by ramnes — APPROVED** [review]

File: .github/workflows/test_3-8.yml
Comment (tych0) [inline .github/workflows/test_3-8.yml:20]: This seems to duplicate a lot of stuff for not a lot of value IMO (CI being too slow is not a big blocker for this project since we switched to github actions). Is there any way we can at least share what packages need to be installed?

File: .github/workflows/ci.yml
Comment (tych0) [inline .github/workflows/ci.yml]: +1 to this, if you make a separate PR I'm happy to take it.

File: .github/workflows/test_3-8.yml
Comment (DorianGray) [reply .github/workflows/test_3-8.yml:20]: unfortunately no, github actions is working on features to make this setup less bloated. This is the recommended way to both increase parallelization and allow re-running of individual segments :/

File: .github/workflows/ci.yml
Comment (DorianGray) [reply .github/workflows/ci.yml]: yeah, we weren't actually doing static analysis on large sections of the code because of these missing libraries.

Comment (DorianGray) [reply .github/workflows/ci.yml]: I can make this a separate PR I thnk

File: .github/workflows/test_3-8.yml
Comment (DorianGray) [reply .github/workflows/test_3-8.yml:20]: related discussion : https://github.community/t/ability-to-rerun-just-a-single-job-in-a-workflow/17234/35

Comment (ramnes) [reply .github/workflows/test_3-8.yml:20]: We can share the package that needs to be installed with a base Docker image I think.

Comment (DorianGray) [reply .github/workflows/test_3-8.yml:20]: That's a good idea, want me to build one? It'll need a special place to be hosted on dockerhub.

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: I guess what I'm wondering is: is there a reason to merge anything like this at all? What does it achieve, and is that worth the additional complexity?

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: (These are all great things to put in commit messages :)

Comment (ramnes) [reply .github/workflows/test_3-8.yml:20]: It is in the commit message actually:

> This breaks out python versions into their own workflows to allow re-running only a single version.

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: I saw that, but that seems like a solution in search of a problem. I currently do not have the need to be able to run the tests individually. Adding lots of copypasta or another point of failure in the form of a dependency on a docker registry is not worth it to solve a problem we don't have.

File: .github/workflows/ci.yml
Comment (DorianGray) [reply .github/workflows/ci.yml]: As I noted in IRC, actually this is in another PR: https://github.com/qtile/qtile/pull/2102

Comment (tych0) [reply .github/workflows/ci.yml]: But that has the composable requirements as well :). Any chance we can get this alone?

File: .github/workflows/test_3-8.yml
Comment (ramnes) [reply .github/workflows/test_3-8.yml:20]: The tests are somehow reliable since the last fixes, but just before that it was a nightmare to have them complete. We're doing fancy stuff with processes and alike, so the likelihood of race problems is pretty high. When it's broken, it's nice to be able to run the tests individually as a last resort while the tests are being fixed.

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: > When it's broken, it's nice to be able to run the tests individually as a last resort while the tests are being fixed.

I'd prefer not to merge stuff for this case. Re-running the tests to get things green is not a habit I think we want to get into. We should fix the tests instead.

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: If that blocks merging things because merging things is painful, so be it. That will encourage more people to look at fixing things :)

File: .github/workflows/ci.yml
Comment (DorianGray) [reply .github/workflows/ci.yml]: done https://github.com/qtile/qtile/pull/2162

File: .github/workflows/test_3-8.yml
Comment (DorianGray) [reply .github/workflows/test_3-8.yml:20]: I'm definitely in the "lets not slow down all development because one person made a mistake" boat.

Comment (ramnes) [reply .github/workflows/test_3-8.yml:20]: @tych0 I agree with you to some degree, but look how long it took for someone to find the problem last time. Not everyone has the time to investigate such obscure errors and yet could help on other subjects meanwhile.

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: It's not about mistakes. If there is clearly a bad commit, we can revert it and move on. The problem is when there *isn't* a clear bad commit, and people just re-run CI until it passes and merge. Everyone gets complacent, and CI becomes less and less reliable as more and more semi-reliable tests are introduced. I am not interested in going down this road, or any patches whose sole motivation is to make it easier to go down this road, especially if they introduce additional complexity.

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: > but look how long it took for someone to find the problem last time.

Yes, but merging more potentially broken code is not a solution. It is a path to madness.

**Comment by ramnes** [comment]
I'm +0 and @tych0 is -1, so let's save everyone's time and close this for now. Thank you for the effort @DorianGray.

File: .github/workflows/test_3-8.yml
Comment (DorianGray) [reply .github/workflows/test_3-8.yml:20]: If your concern is test reliability you might take a look at https://github.com/qtile/qtile/pull/2107, although it's become quite big.

Comment (tych0) [reply .github/workflows/test_3-8.yml:20]: Yeah, I think there's maybe some rebase issues with the most important patch there; I haven't looked too closely at it though.

Comment (DorianGray) [reply .github/workflows/test_3-8.yml:20]: I don't see any...O.o

**Pull request closed without merging**
